src/utils/mod.rs

```diff
@@ -0,0 +1,233 @@
+// SPDX-FileCopyrightText: 2020 Serokell <https://serokell.io/>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+use std::borrow::Cow;
+use std::path::PathBuf;
+
+use merge::Merge;
+
+#[macro_export]
+macro_rules! good_panic {
+    ($($tts:tt)*) => {{
+        error!($($tts)*);
+        std::process::exit(1);
+    }}
+}
+
+pub mod data;
+pub mod deploy;
+pub mod push;
+
+pub struct CmdOverrides {
+    pub ssh_user: Option<String>,
+    pub profile_user: Option<String>,
+    pub ssh_opts: Option<String>,
+    pub fast_connection: Option<bool>,
+    pub auto_rollback: Option<bool>,
+    pub hostname: Option<String>,
+    pub magic_rollback: Option<bool>,
+    pub temp_path: Option<String>,
+    pub confirm_timeout: Option<u16>,
+}
+
+#[derive(PartialEq, Debug)]
+pub struct DeployFlake<'a> {
+    pub repo: &'a str,
+    pub node: Option<&'a str>,
+    pub profile: Option<&'a str>,
+}
+
+pub fn parse_flake(flake: &str) -> DeployFlake {
+    let flake_fragment_start = flake.find('#');
+    let (repo, maybe_fragment) = match flake_fragment_start {
+        Some(s) => (&flake[..s], Some(&flake[s + 1..])),
+        None => (flake, None),
+    };
+
+    let (node, profile) = match maybe_fragment {
+        Some(fragment) => {
+            let fragment_profile_start = fragment.find('.');
+            match fragment_profile_start {
+                Some(s) => (Some(&fragment[..s]), Some(&fragment[s + 1..])),
+                None => (Some(fragment), None),
+            }
+        }
+        None => (None, None),
+    };
+
+    DeployFlake {
+        repo,
+        node,
+        profile,
+    }
+}
+
+#[test]
+fn test_parse_flake() {
+    assert_eq!(
+        parse_flake("../deploy/examples/system#example"),
+        DeployFlake {
+            repo: "../deploy/examples/system",
+            node: Some("example"),
+            profile: None
+        }
+    );
+
+    assert_eq!(
+        parse_flake("../deploy/examples/system#example.system"),
+        DeployFlake {
+            repo: "../deploy/examples/system",
+            node: Some("example"),
+            profile: Some("system")
+        }
+    );
+
+    assert_eq!(
+        parse_flake("../deploy/examples/system"),
+        DeployFlake {
+            repo: "../deploy/examples/system",
+            node: None,
+            profile: None,
+        }
+    );
+}
+
+pub struct DeployData<'a> {
+    pub node_name: &'a str,
+    pub node: &'a data::Node,
+    pub profile_name: &'a str,
+    pub profile: &'a data::Profile,
+
+    pub cmd_overrides: &'a CmdOverrides,
+
+    pub merged_settings: data::GenericSettings,
+}
+
+pub struct DeployDefs<'a> {
+    pub ssh_user: Cow<'a, str>,
+    pub profile_user: Cow<'a, str>,
+    pub profile_path: Cow<'a, str>,
+    pub current_exe: PathBuf,
+    pub sudo: Option<String>,
+}
+
+impl<'a> DeployData<'a> {
+    pub fn defs(&'a self) -> DeployDefs<'a> {
+        let ssh_user: Cow<str> = match self.merged_settings.ssh_user {
```

Comment:
`unwrap_or`

Reply:
More like
```rust
let ssh_user: Cow<str> = self
    .merged_settings
    .ssh_user
    .as_deref()
    .map(|x| x.into())
    .unwrap_or_else(|| whoami::username().into());
```
which I think is a little uglier, and longer